register-schemas.R

```diff
@@ -1,48 +1,123 @@
+#' @param dryRun Should the dryRun field be set to `true`? Defaults to `FALSE`.
+#' @return String containing the JSON request body
+create_body <- function(file, dryRun = FALSE) {
+  stopifnot(dryRun %in% c(TRUE, FALSE))
   paste(
     '{"concreteType": "org.sagebionetworks.repo.model.schema.CreateSchemaRequest", "schema": ',
     paste(readLines(file), collapse = ""),
+    ', "dryRun": ',
+    ifelse(dryRun, 'true', 'false'),
     '}'
   )
 }
 
-register_schema <- function(file) {
+#' Register schema to Synapse
+#'
+#' Given a path to a file containing a JSON schema, registers that schema on
+#' Synapse.
+#'
+#' @param file Path to JSON Schema file
+#' @param dryRun Should the dryRun field be set to `true`? Defaults to `FALSE`.
+#' @return Token object containing the async token used to monitor the job
+register_schema <- function(file, dryRun = FALSE) {
   synRestPOST(
     uri = "/schema/type/create/async/start",
-    body = create_body(file)
+    body = create_body(file, dryRun = dryRun)
   )
 }
 
-# Register terms ---------------------------------------------------------------
-
-term_files <- dir_ls(here("terms"), regexp = "*\\.json", recurse = 1)
+#' Get status of registered schema
+#'
+#' @param token Async token
+#' @return Response object containing schema
+get_registration_status <- function(token) {
+  processing <- TRUE
+  ## Check results of registering schema, retrying if the async job hasn't
+  ## completed yet
+  while (processing) {
+    result <- synRestGET(
+      uri = glue("/schema/type/create/async/get/{token}")
+    )
+    ## synapser doesn't return the status codes unfortunately, so we check the
+    ## response object to determine what to do. If it contains "jobState",
+    ## that's part of the AsynchronousJobStatus, i.e. the async job hasn't
+    ## completed yet.
+    if (!"jobState" %in% names(result)) {
+      processing <- FALSE
+    }
+  }
+  result
+}
 
-## Register each mini-schema
-walk(term_files, register_schema)
+#' Register schema and report on its status
+#'
+#' Registers a schema and then monitors the asynchronous job until it completes.
+#'
+#' @param file Path to JSON Schema file
+#' @param dryRun Should the dryRun field be set to `true`? Defaults to `FALSE`.
+#' @return `TRUE` if schema was successfully registered; `FALSE` otherwise.
+register_and_report <- function(file, dryRun = FALSE) {
+  message(glue("Registering {file} with dryRun = {dryRun}"))
+  ## register_schema() can fail if e.g. schemas contain non-ASCII characters --
+  ## in this case we catch that error and provide the message.
+  token <- try(register_schema(file, dryRun = dryRun), silent = TRUE)
+  if (inherits(token, "try-error")) {
+    message(token)
+    return(FALSE)
+  }
+  ## If we get a 400 status code, synapser will throw an error -- we want to
+  ## catch that and provide the message but not raise an actual R error.
+  status <- try(get_registration_status(token$token), silent = TRUE)
```

Comment:
I was worried about the while loop on line 67, but you're catching possible errors here.